linux-project/Shell.cpp

```diff
@@ -0,0 +1,256 @@
+     *
+     * @param argumentArray The array of C-style strings representing the command and its arguments.
+     */
+
+    int statusCode = execvp(argumentArray[0], argumentArray);
+    if (statusCode == -1) {
+        std::cerr << argumentArray[0] << ": command not found.\n";
+        exit(1);
+    }
+}
+
+void Shell::runParentProcess(char* const* argumentArray, const pid_t& pid, const bool& runInBackground, const std::string& commandLine) {
+    /**
+     * Executes the parent process for an external command.
+     * It waits for the child process to finish if not running in the background.
+     * If running in the background, it adds the background process to the backgroundProcesses_ map.
+     *
+     * @param argumentArray The array of C-style strings representing the command and its arguments.
+     * @param pid The process ID of the child process.
+     * @param runInBackground Flag indicating if the command should run in the background.
+     * @param commandLine The user command line.
+     */
+
+    int returnStatus;
+
+    if (!runInBackground) {
+        if (waitpid(pid, &returnStatus, 0) == -1) {
+            throw std::runtime_error("Wait error.");
+        }
+
+    }
+    else {
+        m_backgroundProcesses[pid] = { commandLine, -1, std::time(0) };
+        std::cout << "Process running in the background with PID: " << pid << std::endl;
+    }
+
+}
+
+void Shell::showBackgroundProcesses() {
+    /**
+     * Displays the information about the background processes currently running.
+     * It prints the process ID, command, status, and additional details (e.g., time running).
+     */
+
+    if (m_backgroundProcesses.size() == 0)
+        std::cout << "No background processes running\n";
+
+    else {
+        for (const auto& [pid, jobInfo] : m_backgroundProcesses) {
+            std::cout << "PID: " << pid 
+                << " | Command: " << jobInfo.command 
+                << " | Status: " << jobInfo.status 
+                << " | time running: " << difftime(std::time(0), jobInfo.startTime) << " seconds\n";
+        
+        }
+    }
+}
+
+void Shell::cleanUpFinishedProcesses() {
+    /**
+     * Cleans up the finished background processes by checking their status using waitpid.
+     * It removes the finished processes from the backgroundProcesses_ map and prints relevant information.
+     */
+
+    for (auto it = m_backgroundProcesses.begin(); it != m_backgroundProcesses.end();) {
+        int status;
+        pid_t result = waitpid(it->first, &status, WNOHANG);
+        
+        if (result == it->first) {
+            std::cout << "process with pid: " << it->first 
+                << " runned by command: " << it->second.command
+                << " has finished with status " << WEXITSTATUS(status)
+                <<". Execution time : " << difftime(std::time(0), it->second.startTime) << " seconds\n";
+            m_backgroundProcesses.erase(it++);
```

Comment:
Erasing the iterator and then incrementing it might be unsafe. This is because when you remove the iterator it gets invalidated. You should use the return value of the `erase` method instead. 
Look at the example here: https://en.cppreference.com/w/cpp/container/map/erase